### DISCUSSION TITLE
async in useState

### DISCUSSION THREAD

**Discussion by Gringo675** [nuxt/nuxt#17654] [Questions] [answered]
Hi guys!
If there is a way to get asynchronies data within useState init function? Something like

```
const counter = useState('counter', async () => {
	const result = await someAsyncFunc()
	return result
})

```

All my attempts doing this failed. Even tried assign value after initialization, like this

```
const counter = useState('counter', () => {
	initCounter()
	return 0
})

async function initCounter() {
	const counter = useState('counter)
	counter.value = await someAsyncFunc()
}

```

It doesn't work, too :) So, is it impossible?

**Comment by danielroe** [comment, marked as answer]
For this, you can use `useAsyncData`:

```js
const { data: counter } = await useAsyncData('counter', async () => {
  const result = await someAsyncFunc()
  return result
})
```

**Reply by Gringo675** [reply]
Sorry, think you have misprint in your code, there is no useAsyncData. Do you mean apply useAsyncData insead of useState?

**Reply by danielroe** [reply]
Exactly! Updated it; my apologies...

**Reply by Gringo675** [reply]
Thanks, but can't understand. Know that I could take asynchronous data with `$fetch` or `useAsyncData`, but the question was how to do it within useState init function, e.g. do it global.
Of course I can declare a variable with useState, and then fetch async data, but if there is a way to do it in one place, when initializing a variable?

**Reply by danielroe** [reply]
useAsyncData is global, based on the key you set.

**Reply by Gringo675** [reply]
Global? Then how get access to it? Sorry, maybe it is stupid question, but I can't understand it or find in documentation.
I've tried:

```
const { data: counter } = await useAsyncData('counter', () => someAsyncFunc());  // or your variant

async function someAsyncFunc() {
  let promise = new Promise((resolve, reject) => {
    setTimeout(() => resolve(777), 3000)
  });
  return await promise;
}

```


And then get it in another component:
`const counter = useState('counter')`
`const counter = useAsyncData('counter')`

In first case it is undefined, in second - error.

**Comment by Gringo675** [comment]
Think I've figured it out.

In one component we can get some async data:
`const {data:counter} = useAsyncData('testAsyncDataKey', () => someAsyncFunc())`

And now we can get this variable in another component through this:
`const {data:counter} = useAsyncData('testAsyncDataKey', ()=>{})`

Is it correct way to do so?

**Reply by danielroe** [reply]
I would recommend instead exporting your own composable, e.g.:

```js
async function someAsyncFunc() {
  let promise = new Promise((resolve, reject) => {
    setTimeout(() => resolve(777), 3000)
  });
  return await promise;
}

export const useCounter = () => useAsyncData('counter', someAsyncFunc);  // or your variant
```

That way it always has the same initialiser function, so you don't need to worry about the _order_ in which you call it, or have to worry about syncing the key between times you call it.

**Reply by Gringo675** [reply]
Oh, yes, great. Thanks a lot!

**Comment by lukas-schaetzle** [comment]
The marked solution doesn't seem to work. When I have two pages A and B which shall display the same state (for example a timestamp created at the state's initialization) I can use `useState` like this to get the desired effect, and navigating between the two pages does not reinitialize the state.
```js
export const useTime = () => useState('time', () => (new Date()).getSeconds())

// used in page like this: const time= useTime()
```

However, when using an async intialization function with `useAsyncData` the state is reinitialized after navigating to the other page.
```js
export const useTime = () => useAsyncData('time', async () => {
  await new Promise((resolve) => {
    setTimeout(() => resolve('resolved'), 1000)
  })
  return new Date().getSeconds()
})

// used in page like this: const time= await useTime()
```

Tested with Nuxt 3.3.2

**Reply by lukas-schaetzle** [reply]
Turns out, the shown examples in the marked solution are a bit irritating because they do not show actually re-using the old state set by `useAsyncState`. By default, `useAsyncData` has its `immediate` option set to `true`, thus it refreshes every time it is called. If you want something like `useState` with an async init function, you can use a composable like this:

```js
export const useTime = async () => {
  const key = 'time'
  const { data: time } = useNuxtData(key);
  if (time.value) {
    return time;
  }
  const { data } = await useAsyncData(key, async () => {
    await new Promise((resolve) => {
      setTimeout(() => resolve('resolved'), 1000);
    });
    return new Date().getSeconds();
  });
  return data;
};
```
Also see https://github.com/nuxt/nuxt/issues/19972#issuecomment-1488270660

**Reply by Gringo675** [reply]
Yes, or little shorter:
````
useAsyncData('key', () => {
const oldValue = useNuxtData('key').data.value
if (oldValue) return oldValue
...
}
````

**Comment by Gringo675** [comment]
Yes, useAsyncData behavior changed for the last year. Since rc14, I think. Now when reapplying it flashes updating of data. Developers think it's better, don't know. Especially it's annoying with lazy option, when composable returns "old" data with pending = false, and yet starts asynchronously updating data.
